addon/v1alpha1/types_managedclusteraddon.go

```diff
@@ -0,0 +1,134 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Progressing",type=string,JSONPath=`.status.conditions[?(@.type=="Progressing")].status`
+// +kubebuilder:printcolumn:name="Available",type=string,JSONPath=`.status.conditions[?(@.type=="Available")].status`
+
+// ManagedClusterAddOn is the Custom Resource object which holds the current state
+// of an add-on. This object is used by add-on operators to convey their state.
+// The resource should be created in the ManagedCluster namespace.
+type ManagedClusterAddOn struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// spec holds configuration that could apply to any operator.
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ManagedClusterAddOnSpec `json:"spec"`
+
+	// status holds the information about the state of an operator.  It is consistent with status information across
+	// the Kubernetes ecosystem.
+	// +optional
+	Status ManagedClusterAddOnStatus `json:"status"`
+}
+
+// ManagedClusterAddOnSpec is empty for now.
+type ManagedClusterAddOnSpec struct {
+}
+
+// ManagedClusterAddOnStatus provides information about the status of the operator.
+// +k8s:deepcopy-gen=true
+type ManagedClusterAddOnStatus struct {
+	// conditions describe the state of the managed and monitored components for the operator.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +optional
+	Conditions []AddOnStatusCondition `json:"conditions,omitempty"  patchStrategy:"merge" patchMergeKey:"type"`
```

Comment:
can you use `metav1.Condition` from kube 1.19?

Reply:
currently the api project is not in kube 1.19 yet, I think @qiujian16 is working on this in a separate PR. will update the spec once that's in

Reply:
We will need to have support for this API going back to Kube 1.17 at least.

Reply:
@mdelder this will not prevent backward compatibility issue